src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/collectors/NodeStatsAllShardsMetricsCollector.java

```diff
@@ -48,17 +49,25 @@
             NodeStatsAllShardsMetricsCollector.class).samplingInterval;
     private static final int KEYS_PATH_LENGTH = 2;
     private static final Logger LOG = LogManager.getLogger(NodeStatsAllShardsMetricsCollector.class);
-    private HashMap<String, IndexShard> currentShards;
+    private HashMap<ShardId, IndexShard> currentShards;
+    HashMap<ShardId, ShardStats> currentPerShardStats;
+    HashMap<ShardId, ShardStats> prevPerShardStats;
     private final PerformanceAnalyzerController controller;
 
 
     public NodeStatsAllShardsMetricsCollector(final PerformanceAnalyzerController controller) {
         super(SAMPLING_TIME_INTERVAL, "NodeStatsMetrics");
         currentShards = new HashMap<>();
+        prevPerShardStats = new HashMap<>();
+        currentPerShardStats = new HashMap<>();
         this.controller = controller;
     }
 
     private void populateCurrentShards() {
+        if (currentShards.size() != 0) {
+            prevPerShardStats = currentPerShardStats;
+            currentPerShardStats.clear();
```

Comment:
In java, all object are passed by reference. So, when we do `prevPerShardStats = currentPerShardStats;`, the `prevPerShardStats` gets a reference to the actual value on heap, which we are clearing with `currentPerShardStats.clear();`

Instead we need to perform either a Deep Copy on `currentPerShardStats` or Create a new HashMap from `currentPerShardStats`

Reply:
used putAll() which creates a deep-copy of the object.